fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/bitmap/ComposedBitmapIndexCursor.java

```diff
@@ -0,0 +1,165 @@
+         */
+        @Nullable
+        byte[] compose(@Nonnull List<byte[]> bitmaps, int size);
+    }
+
+    protected ComposedBitmapIndexCursor(@Nonnull List<MergeCursorState<IndexEntry>> cursorStates, @Nullable FDBStoreTimer timer, @Nonnull Composer composer) {
+        super(cursorStates, timer);
+        this.composer = composer;
+    }
+
+    @Nonnull
+    @Override
+    protected CompletableFuture<List<MergeCursorState<IndexEntry>>> computeNextResultStates() {
+        final List<MergeCursorState<IndexEntry>> cursorStates = getCursorStates();
+        return whenAll(cursorStates).thenApply(vignore -> {
+            boolean anyHasNext = false;
+            for (MergeCursorState<IndexEntry> cursorState : cursorStates) {
+                if (cursorState.getResult().hasNext()) {
+                    anyHasNext = true;
+                } else if (cursorState.getResult().getNoNextReason().isLimitReached()) {
+                    // Stop if any has reached limit.
+                    return Collections.emptyList();
+                }
+            }
+            if (anyHasNext) {
+                final List<MergeCursorState<IndexEntry>> resultStates = new ArrayList<>();
+                long nextPosition = Long.MAX_VALUE;
+                for (MergeCursorState<IndexEntry> cursorState : cursorStates) {
+                    if (cursorState.getResult().hasNext()) {
+                        final IndexEntry indexEntry = cursorState.getResult().get();
+                        final Tuple indexKey = indexEntry.getKey();
+                        final long position = indexKey.getLong(indexKey.size() - 1);
+                        if (nextPosition > position) {
+                            resultStates.clear();
+                            nextPosition = position;
+                        }
+                        if (nextPosition == position) {
+                            resultStates.add(cursorState);
+                        }
+                    }
+                }
+                return resultStates;
+            } else {
+                return Collections.emptyList();
+            }
+        });
+    }
+
+    @Nonnull
+    @Override
+    protected IndexEntry getNextResult(@Nonnull List<MergeCursorState<IndexEntry>> resultStates) {
+        final List<MergeCursorState<IndexEntry>> cursorStates = getCursorStates();
+        final IndexEntry firstEntry = resultStates.get(0).getResult().get();
+        final int size = firstEntry.getValue().getBytes(0).length;
+        final List<byte[]> bitmaps = new ArrayList<>(cursorStates.size());
+        for (MergeCursorState<IndexEntry> cursorState : cursorStates) {
+            if (resultStates.contains(cursorState)) {
+                byte[] bitmap = cursorState.getResult().get().getValue().getBytes(0);
+                if (bitmap.length != size) {
+                    throw new RecordCoreException("Index bitmaps are not all the same size");
```

Comment:
Should it be part of the planning that the parameter for the size needs to be the same? It seems like the "ideal" version would figure it out then, but maybe it's sufficient for the time being for it to be an error here.